bucket_map/src/index_entry.rs

```diff
@@ -83,21 +83,38 @@ impl BucketOccupied for BucketWithHeader {
 #[derive(Debug)]
 pub struct IndexBucketUsingBitVecBits<T: PartialEq + 'static> {
     /// 2 bits per entry that represent a 4 state enum tag
-    pub enum_tag: BitVec,
+    pub enum_tag_first_bit: BitVec,
+    /// second will be empty in all healthy cases because in real use, we only use enum values 0 and 1
+    pub enum_tag_second_bit: BitVec,
     /// number of elements allocated
     capacity: u64,
     _phantom: PhantomData<&'static T>,
 }
 
 impl<T: Copy + PartialEq + 'static> IndexBucketUsingBitVecBits<T> {
     /// set the 2 bits (first and second) in `enum_tag`
-    fn set_bits(&mut self, ix: u64, first: bool, second: bool) {
-        self.enum_tag.set(ix * 2, first);
-        self.enum_tag.set(ix * 2 + 1, second);
+    pub(crate) fn set_bits(&mut self, ix: u64, first: bool, second: bool) {
```

Comment:
nit: how about moving !second check out of is_empty() check, which saves the is_empty check when !second.

```
        if !second { return; }

        if self.enum_tag_second_bit.is_empty() {
               // The first time someone sets the second bit, we have to allocate and check it.
               self.enum_tag_second_bit = BitVec::new_fill(false, self.capacity);
        }
        self.enum_tag_second_bit.set(ix, second);
```

Reply:
if we do the suggestion, then we'd never UNset the second bit if it had previously been set for an entry. We have to do it like this, I think.

calling set_bits with 2=true would allocate and set second bit to true. calling set_bits with2=false would then break out early and leave the second bit set to true.

Reply:
You are right. Thanks for explaining.